src/tokens/ERC1155.sol

```diff
@@ -0,0 +1,129 @@
+abstract contract ERC1155 is ERC1155TokenReceiver {
+
+  /*///////////////////////////////////////////////////////////////
+                            ERC1155 STORAGE
+  //////////////////////////////////////////////////////////////*/
+
+  /// @dev onReceive function signatures
+  bytes4 constant internal ERC1155_RECEIVED_VALUE = 0xf23a6e61;
+  bytes4 constant internal ERC1155_BATCH_RECEIVED_VALUE = 0xbc197c81;
+
+  mapping (address => mapping(uint256 => uint256)) internal balances;
+
+  mapping (address => mapping(address => bool)) internal operators;
+
+  /*///////////////////////////////////////////////////////////////
+                                EVENTS
+  //////////////////////////////////////////////////////////////*/
+
+  event TransferSingle(address indexed operator, address indexed from, address indexed to, uint256 id, uint256 amount);
+
+  event TransferBatch(address indexed operator, address indexed _from, address indexed _to, uint256[] ids, uint256[] amounts);
+
+  event ApprovalForAll(address indexed owner, address indexed operator, bool approved);
+
+  /*///////////////////////////////////////////////////////////////
+                    Public Transfer Functions
+  //////////////////////////////////////////////////////////////*/
+
+  function safeTransferFrom(address from, address to, uint256 id, uint256 amount, bytes memory data) public override {
+    require((msg.sender == from) || isApprovedForAll(from, msg.sender), "INVALID_OPERATOR");
+    require(to != address(0), "INVALID_RECIPIENT");
+    // require(amount <= balances[from][id]) is not necessary since checked with safemath operations
+
+    _safeTransferFrom(from, to, id, amount);
+    _callonERC1155Received(from, to, id, amount, gasleft(), data);
+  }
+
+  function safeBatchTransferFrom(address from, address to, uint256[] memory ids, uint256[] memory amounts, bytes memory data) public override {
+    require((msg.sender == from) || isApprovedForAll(from, msg.sender), "INVALID_OPERATOR");
+    require(to != address(0), "INVALID_RECIPIENT");
+
+    _safeBatchTransferFrom(from, to, ids, amounts);
+    _callonERC1155BatchReceived(from, to, ids, amounts, gasleft(), data);
+  }
+
+  /*///////////////////////////////////////////////////////////////
+                    Internal Transfer Functions
+  //////////////////////////////////////////////////////////////*/
+
+  function _safeTransferFrom(address from, address to, uint256 id, uint256 amount) internal {
+    balances[from][id] = balances[from][id] - amount;
+    balances[to][id] = balances[to][id] + amount;
+
+    emit TransferSingle(msg.sender, from, to, id, amount);
+  }
+
+  function _callonERC1155Received(address from, address to, uint256 id, uint256 amount, uint256 gasLimit, bytes memory data) internal {
+    if (to.code.length > 0) {
+      bytes4 retval = ERC1155TokenReceiver(to).onERC1155Received{gas: gasLimit}(msg.sender, from, id, amount, data);
```

Comment:
What should I do here to use the onERC1155Received param on `to`?

Reply:
that looks fine— what am i missing?